man/dbigen.1
.TH DBIGEN 1 " 4 April 1996"
.FB
.SH NAME
dbigen \- generate a \fIFirstBase\fP index
.SH SYNOPSIS
.B dbigen
[\fB-b\fP] [\fB-c\fP] [\fB-f\fP] [\fB-n\fP] [\fB-r\fP] [\fB-y\fP] [\fB-B\fP]
[\fB-d\fP \fIdatabase\fP]
[\fB-i\fP \fIindex\fP] [\fB-x\fP \fIxref-index\fP]
[\fB-btree\fP]
.SH DESCRIPTION
.I Dbigen
generates a \fIFirstBase\fP Btree+ index or a Normal or Flat index using an
index generator dictionary defined via
\fIdbdind\fP(1).
This index dictionary is traversed during
index generation to selectively
determine which records are to be included in the
index being built.
.PP
The default type of index defined via \fIdbdind\fP(1) is a Btree+ index.
However, by using the \fB-btree\fP command line flag,
\fIdbigen\fP(1) can be forced to create a Btree+ index even when the
dictionary represents a normal index.
.PP
For more information on the selection criteria in the index dictionary,
see the discussion in \fIdbdind\fP(1).
.PP
Note that \fIdbigen\fP(1) can detect when an index is already up to date.
In this case, an index will not be generated. Override this with the
\fB-f\fP (force) option.
.PP
The \fB-B\fP flag will enable the fastest throughput of records from the
database. However, since the database map is not followed when this option
is used, the database must be clean (as in \fIdbclean\fP(1)),
with no free list entries.
(See \fIdatabase\fP(4)). If this flag can be used, the index
generation is speeded up a great deal by internal record buffering.
Note that no cross reference index can be used with record buffering, and
that record locking is disabled.
.PP
The \fB-b\fP
flag will run the \fIFirstBase\fP tool in batch mode, suitable for shell
scripts.
This mode produces no output on the terminal, thus can also be used
when running the process in the background.
The \fB-y\fP flag will run the tool at the terminal without pausing to ask
the user if existing files can be destroyed and without displaying
a 'picture' of the requested task.
.PP
The \fB-c\fP flag is for \fIcase-insensitive\fP searching. All tests for
string matches, both normal and regular expressions, are done using
UPPERCASE so that matches are case insensitive.
.PP
The \fB-f\fP flag is used to force \fIdbigen\fP(1) to run \(em even if it
thinks it does not need to. Use of this flag is equivalent to
runing \fIrmidx\fP(8) before running \fIdbigen\fP(1).
.PP
When using the regular expression pattern matching mechanism, some
machines allow the meta character '.' to match a newline character and some
do not. Typically, BSD UNIX allows this match and UNIX System V does not.
The \fB-n\fP flag can be used to force \fIFirstBase\fP running on System V UNIX
to map newlines to blanks
so that the '.' within regular expressions \fIwill\fP match a newline.
.PP
The \fB-r\fP
flag will cause dbigen to report
the actual number of records selected \(em written to the standard output \(em 
even when the process is run in batch mode.
.PP
\fIDbigen\fP(1)
parses the index dictionary and displays an image of what it
will generate. The user can abort at this point if necessary.
.PP
The \fB-x\fP command line
option can be used to enter the name of an existing \fIFirstBase\fP index to
use
in building the new index. If this existing index is to be used to process
the database to build the target index, and if this existing index is out
of date, an attempt is made to automatically update it first.
.PP
Thus, when using the \fB-x\fP option,
the existing index becomes a cross reference index.
Without this option, the entire \fIFirstBase\fP database is used to select
and build the requested index.
.PP
After successful index generation, the index is sorted automatically.
The standard UNIX sort facility, \fIsort\fP(1), is used for this task.
.PP
.I Note:
date field indexes are reformatted before being written to the index
to enable the system sort to work properly. The internal representation of
indexed dates is YYMMDD (year/month/day) instead of the normal MMDDYY.
This is meant to be transparent to users.
.PP
All NEWLINE characters in the sort by strings are converted to BLANKs.
.SH EXAMPLES
To run a simple interactive dbigen, use
.sp 1
.ti +5
dbigen -d dbase -i index
.sp 1
To run this in batchmode, providing a second index to work from, use:
.sp 1
.ti +5
dbigen -b -d dbase -i index -x xindex
.SH FILES
.PD 0
.TP 10
dbase
default name for the \fIFirstBase\fP database.
.TP 10
index
default name for the \fIFirstBase\fP index.
.TP 10
*.idicti
dictionary for index generator.
.TP 10
*.idict
index header information.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.idx
\fIFirstBase\fP index.
.TP 10
*.map
record map of \fIFirstBase\fP database.
.PD
.SH SEE ALSO
dbdind(1), dbedit(1), sort(1), dbfilter(1), firstbase(5), rmidx(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br
.SH DIAGNOSTICS
If the index file already exists, \fIdbigen\fP(1)
will pause and ask permission to overwrite it, unless batch mode or yes mode
is indicated, in which case dbigen will assume it is ok to overwrite
the existing index.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbigen/dbigen.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbigen.c,v 9.1 2001/02/16 19:43:51 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Dbigen_sid[] = "@(#) $Id: dbigen.c,v 9.1 2001/02/16 19:43:51 john Exp $";
#endif

/*
 *  dbigen.c - index generator (uses idicti)
 */

#include <fb.h>
#include <fb_vars.h>
#include <igen.h>
#include <sys/types.h>
#include <sys/stat.h>

static char *MSG1 = "Index is Current and Up to Date!";

fb_database *hp;

/*
 *  dbigen - main driver
 */

fb_field *by[FB_MAXBY];			/* sortby fields */
struct self *top;			/* top of the select tree */
int rflag = 0;				/* report sorted by flag */
int cflag = 0;				/* case-insensative flag */
int fflag = 0;				/* force flag - uncond. regen */
int nflag = 0;				/* newline to blank mapping flag */
int Bflag = 0;				/* Blocking factor flag */
int btree = 0;				/* gen a Btree index when done */

static getindex();

   main(argc, argv)
      int argc;
      char *argv[];
   
      {
         struct self *stree();
         char gidict[FB_MAXNAME], gindex[FB_MAXNAME], bindex[FB_MAXNAME];
	 int i;
	 struct stat dbuf, xbuf, ibuf;
   
         fb_getargs(argc, argv, FB_ALLOCDB);
	 hp = cdb_db;
	 if (fb_testargs(argc, argv, "-f") > 0)
	    fflag = 1;
	 /* if the index is current, do not bother with igen */
	 dbuf.st_mtime = xbuf.st_mtime = ibuf.st_mtime = 0;
	 sprintf(gidict, "%si", hp->idict);	/* borrow the gidict area */
	 if (!fflag && stat(hp->dbase, &dbuf) == 0	&&
	     stat(hp->dindex, &xbuf) == 0	&&
	     stat(gidict, &ibuf) == 0 		&&
	     access(hp->idict, 0) == 0){	/* files must exist */
	    if (xbuf.st_mtime > ibuf.st_mtime &&
		xbuf.st_mtime > dbuf.st_mtime)
	       if (fb_checktotals(hp) == FB_AOK){
		  if (!cdb_batchmode && !cdb_yesmode)
		     fb_serror(FB_MESSAGE, MSG1, NIL);
		  fb_ender();			/* rah rah - up to date */
		  }
	    }
	 
	 /* deep sigh. run it. */
	 if (fb_testargs(argc, argv, "-B") > 0)
	    Bflag = 1;
	 strcpy(gidict, hp->idict);
	 strcpy(gindex, hp->dindex);
	 fb_basename(bindex, hp->dindex);
	 fb_opendb(cdb_db, READ, FB_NOINDEX, 0);
	 fb_scrhdr(hp, bindex); fb_infoline();
	 for (i = 0; i < hp->nfields; i++)
	    fb_nounders(cdb_keymap[i]);
         if ((top = stree(cdb_keymap, hp, by)) == NULL){
	    sprintf(gidict, "%si", hp->idict);
	    fb_xerror(FB_BAD_DICT, gidict, NIL);
	    }
	 if (getindex(argc, argv, gindex) == FB_ABORT)
	    fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	 if (!cdb_batchmode && !cdb_yesmode){
            itrace(top, by, 0);
	    fb_infoline();
            if (fb_mustbe('y',"If accurate, enter 'y' to continue:",
                  cdb_t_lines, 1) != FB_AOK)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    fb_move(4, 1), fb_clrtobot(), fb_infoline();
	    }
	 if (!cdb_batchmode){
	    if (!cflag)
               fb_scrstat("Selecting");
	    else
               fb_scrstat("Selecting-CaseIns");
	    FB_XWAIT();
	    fb_gcounter(0L);
	    }
	 if (cdb_yesmode)
	    sleep(1);				/* some are too fast */
         if (mkidx(gidict, gindex) == FB_ERROR)
            fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	 fb_closedb(cdb_db);
         fb_ender();
      }

/* 
 *  getindex - get index from which the generated index will made from 
 *    the default is to use the actual fb_database itself 
 */
 
   static getindex(argc, argv, gindex)
      char *gindex, *argv[];
      int argc;
      
      {
         char file[FB_MAXNAME], err[FB_MAXNAME];
	 int st, xflag;
	 
	 hp->dindex = hp->idict = NULL;
	 file[0] = NULL;
	 if (fb_testargs(argc, argv, "-r") > 0)
	    rflag = 1;
	 if (fb_testargs(argc, argv, "-c") > 0)
	    cflag = 1;
	 if (fb_testargs(argc, argv, "-n") > 0)
	    nflag = 1;
	 if ((xflag = fb_testargs(argc, argv, "-x")) > 0){
	    if (argc > xflag)
	       strcpy(file, argv[xflag + 1]);
	    else if (cdb_batchmode || cdb_yesmode)	/* cannot be -x as last arg */
	       usage();
	    }
	 if (fb_testargs(argc, argv, "-btree") > 0)
	    btree = 1;
	 if (!xflag)				/* if no -x no ask */
	    return(FB_AOK);
	 for (;;){
	    if (!cdb_batchmode && !xflag){
	       fb_move(4, 1); fb_clrtobot();
	       fb_move(10,10);
	       fb_printw("Enter Name of EXISTING Index to Work From...");
	       fb_move(11,10);
	       fb_printw("Or <RETURN> to Force Use of Entire DataBase:");
	       fb_infoline();
	       fb_scrhlp(SYSMSG[S_CHOICE_MSG1]);
	       st = fb_input(13, 10, 44, 0, FB_ALPHA, file, FB_ECHO, FB_NOEND, FB_CONFIRM);
	       fb_scrhlp(NIL);
	       if (st == FB_DEFAULT || st == FB_ABORT)
		  return(st);
	       }
	    fb_trim(file);
            fb_dbargs(NIL, file, hp);
	    if (equal(hp->dindex, gindex)){
	       if (cdb_batchmode)
	          usage();
	       fb_serror(FB_MESSAGE, SYSMSG[S_ILLEGAL], SYSMSG[S_INVALID]);
	       }
	    else if (fb_geti_dict(READWRITE, hp) == FB_ERROR)
	       fb_xerror(FB_BAD_DICT, hp->idict, NIL);
	    else if (fb_geti_data(READWRITE, hp) == FB_ERROR)
	       fb_xerror(FB_BAD_DATA, hp->dindex, NIL);
	    break;
	    }
	 if (!cdb_batchmode && !cdb_yesmode){
	    sprintf(err, "Using EXISTING index %s to generate %s",
	       hp->dindex,gindex);
	    fb_serror(FB_MESSAGE, err, NIL);
	    }
	 return(FB_AOK);
      }

/* 
 * usage message 
 */

   usage()
      {
         fb_xerror(FB_MESSAGE, "usage: dbigen [-d dbase] [-i index]",
	    "[-y] [-b [-x xref_index]]");
      }
